subprojects/core/src/main/java/org/gradle/api/tasks/SourceTask.java

```diff
@@ -84,8 +94,28 @@ public void setSource(FileTree source) {
      * @param source The source.
      */
     public void setSource(Object source) {
+        final List<Object> oldSources = new ArrayList<Object>(this.source);
         this.source.clear();
-        this.source.add(source);
+        unpackUnionFileCollectionTo(source, new Consumer<Object>() {
+            @Override
+            public void accept(Object newSource) {
+                if (newSource == sourceProperty) {
+                    SourceTask.this.source.addAll(oldSources);
+                } else {
+                    SourceTask.this.source.add(newSource);
+                }
+            }
+        });
+    }
+
+    private static void unpackUnionFileCollectionTo(Object sources, Consumer<Object> consumer) {
```

Comment:
This `unpackUnionFileCollection` seems more like something we should support on `FileCollection` to prevent cycles. I'm leaning toward we should probably not make the change. The tests should stay.